drivers/g3plc.c

```diff
@@ -0,0 +1,889 @@
+	}
+
+	return g3plc_test_run(g3plc_priv.netif, test);
+}
+
+
+static void g3plc_msg_thread(void *arg)
+{
+	msg_t msg = { 0 };
+	msg_rid_t rid;
+	struct g3adp_ctl *ctl;
+
+	for (;;) {
+		if (msgRecv(g3plc_priv.port, &msg, &rid) < 0)
+			continue;
+
+		switch (msg.type) {
+			case mtDevCtl:
+				ctl = (struct g3adp_ctl *)msg.i.raw;
+				if (ctl->type == g3adp_test) {
+					DEBUG_LOG("g3adp: devctl test %s\n", ctl->name);
+					msg.o.err = g3plc_test_cmd(ctl->name);
+				}
+				else if (ctl->type == g3adp_get) {
+					msg.o.err = g3plc_get_val(ctl->name);
+				}
+
+			default:
+				break;
+		}
+
+		msgRespond(g3plc_priv.port, &msg, rid);
+	}
+}
+
+#include "shmem.h"
+#define BAREMETAL_LOG_PASS 0
+#define SAP_SHMEM          1
+
+#if BAREMETAL_LOG_PASS
+#define BAREMETAL_LOG_SERIAL 7
+#endif
+#define SAP_SHMEM_SERIAL 6
+#define DEBUG_RPC_SERIAL (0x3)
+#define DEBUG_RPC_IDX    (0x0)
+
+static shmem_init_locals(void)
+{
+	int res;
+	shmem_range_t shmem_range;
+	res = shmem_init(&shmem_range);
+	if (res < 0) {
+		ERROR_LOG("shmem_init() failed: %d", res);
+		return -1;
+	}
+	res = 0;
+
+	res = ps_serial_init_from_buf_master(SAP_SHMEM_SERIAL, (uint8_t *)shmem_range.start + res, shmem_range.size - res, 1024, 1024);
+	if (res < 0) {
+		ERROR_LOG("ps_serial_init_from_buf_master(SAP) failed: %d", res);
+		return -1;
+	}
+
+	DEBUG_LOG("shmem: sap buffer: start=%p, size=%zu", (uint8_t *)shmem_range.start, shmem_range.size);
+
+	res = shmem_notify_cpu(&shmem_range);
+	if (res < 0) {
+		ERROR_LOG("shmem_notify_cpu() failed: %d", res);
+		return -1;
+	}
+}
+
+static int sap_setup(ps_g3_sap_t *sap)
+{
+	ps_g3_sap_callbacks_t clbk_adp = {
+		.mcps_data_request = ps_g3_sap_ser_mcps_data_request,
+		.mcps_data_confirm = mcps_data_confirm,
+		.mcps_data_indication = mcps_data_indication,
+		.mlme_reset_request = ps_g3_sap_ser_mlme_reset_request,
+		.mlme_reset_confirm = mlme_reset_confirm,
+		.mlme_scan_request = ps_g3_sap_ser_mlme_scan_request,
+		.mlme_beacon_notify_indication = mlme_beacon_notify_indication,
+		.mlme_scan_confirm = mlme_scan_confirm,
+		.mlme_set_request = ps_g3_sap_ser_mlme_set_request,
+		.mlme_set_confirm = mlme_set_confirm,
+		.mlme_get_request = ps_g3_sap_ser_mlme_get_request,
+		.mlme_get_confirm = mlme_get_confirm,
+		.mlme_start_request = ps_g3_sap_ser_mlme_start_request,
+		.mlme_start_confirm = mlme_start_confirm,
+		/* .mlme_comm_status_indication = mlme_comm_status_indication, */
+	};
+
+#define SAP_UART_MODE    0
+#define SAP_PHOENIX_MODE 1
+#define SAP_SHMEM_MODE   2
+#define SAP_MODE         SAP_SHMEM_MODE
+	shmem_init_locals();
+
+#if SAP_MODE == SAP_UART_MODE
+	ps_g3_sap_dl_uart_cfg_t sap_dl_cfg = {
+		.devname = "/dev/uart3",
+	};
+	if (ps_g3_sap_init(sap, &ps_g3_sap_dl_uart, &sap_dl_cfg) < 0) {
+		printf("ps_g3_sap_init() unable to initialize sap layer");
+		return -1;
+	}
+#elif SAP_MODE == SAP_PHOENIX_MODE
+	if (ps_g3_sap_init(sap, &ps_g3_sap_dl_msgcli, NULL) < 0)
+		return -1;
+#elif SAP_UART_MODE == SAP_SHMEM_MODE
```

Comment:
shouldn't this be:
```suggestion
#elif SAP_MODE == SAP_SHMEM_MODE
```
? 

if not - add comment explaining why